src/Sarif.UnitTests/Visitors/ReformattingVisitorTests.cs

```diff
@@ -0,0 +1,60 @@
+﻿// Copyright (c) Microsoft. All rights reserved. Licensed under the MIT        
+// license. See LICENSE file in the project root for full license information. 
+
+using System;
+using FluentAssertions;
+using Xunit;
+
+namespace Microsoft.CodeAnalysis.Sarif.Visitors
+{
+    public class ReformattingVisitorTests
+    {
+        //                                  0          10          20 
+        //                                  01234567 8 901234 5 678901 2 345
+        private static string s_testText = "line1\r\nline2\r\nline3\r\nline4";
+
+        private static Tuple<Region, string>[] s_testCases = new Tuple<Region, string>[]
+        {
+            // Regions specified only by start line
+            new Tuple<Region, string>(new Region() { StartLine = 1 }, "line1"),
+            new Tuple<Region, string>(new Region() { StartLine = 2 }, "line2"),
+            new Tuple<Region, string>(new Region() { StartLine = 3 }, "line3"),
+            new Tuple<Region, string>(new Region() { StartLine = 4 }, "line4"),
+            // Multiline regions, should only return first line
+            new Tuple<Region, string>(new Region() { StartLine = 1, EndLine = 4 }, "line1"),
+            new Tuple<Region, string>(new Region() { StartLine = 2, EndLine = 3 }, "line2"),
+            new Tuple<Region, string>(new Region() { StartLine = 3, EndLine = 4 }, "line3"),
+            new Tuple<Region, string>(new Region() { StartLine = 4, EndLine = 4 }, "line4"),
+            // Char offsets. Should return full line, no matter what subset of line is referenced
+            new Tuple<Region, string>(new Region() { CharOffset = 0, CharLength = 2 }, "line1"),
+            new Tuple<Region, string>(new Region() { CharOffset = 9, CharLength = 1 }, "line2"),
+            new Tuple<Region, string>(new Region() { CharOffset = 16, CharLength = 2 }, "line3"),
+            new Tuple<Region, string>(new Region() { CharOffset = 22, CharLength = 4 }, "line4"),
+            // Binary regions. No return value expected.
+            new Tuple<Region, string>(new Region() { ByteOffset = 20, ByteLength = 5 }, null),
+            // No explicit values is implicitly a binary region with insertion point at beginning of file
```

Comment:
I guess that's right, although the spec doesn't say it explicitly. The spec says that every valid text region includes either `startLine` or `charOffset`, so this isn't a text region. And in a binary region, `length` and `offset` are both allowed to default to 0. Huh.

Reply:
We should consider requiring an explicit indicator of the kind of a region. This means that one of startLine, charOffset or byteOffset must be explicit. Everything else can be used as a default. This would restrict the # of properties that need to be nullable types.

---
In reply to: [202097770](https://github.com/Microsoft/sarif-sdk/pull/942#discussion_r202097770) [](ancestors = 202097770)

Reply:
You may recall, I once proposed that a region have text region and binary region sub-objects:
```
region: {
  text: {
    "startLine": 42
  },
  "binary": {
  }
}
```

Reply:
you did. this doesn't resolve all the strangeness, however, due to Jim's suggestion that we implicitly allow a region to start with a specific column and include all the remaining line text. to do this, we need a nullable charLength

[https://github.com/oasis-tcs/sarif-spec/issues/201](https://github.com/oasis-tcs/sarif-spec/issues/201) 

---
In reply to: [202409592](https://github.com/Microsoft/sarif-sdk/pull/942#discussion_r202409592) [](ancestors = 202409592)

Reply:
You're right. The same is true of your suggestion to have a region type indicator:
```
# charLength and endColumn missing; region extends to end of line.
"region": {
  "type": "text",
  "startLine": 5
},

# Insertion point.
"region": {
  "type": "text",
  "startLine": 5,
  "charLength": 0
}
```
... and again, the SDK can't tell the difference unless `Region.CharLength` is nullable.